compositor_pipeline/src/pipeline/output/whip.rs

```diff
@@ -0,0 +1,246 @@
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<Self, OutputInitError> {
+        let payloader = Payloader::new(options.video, options.audio);
+        let packet_stream = PacketStream::new(packets_receiver, payloader, 1400);
+        let should_close = Arc::new(AtomicBool::new(false));
+        let (init_confirmation_sender, mut init_confirmation_receiver) =
+            oneshot::channel::<Result<(), WhipError>>();
+
+        let whip_ctx = WhipCtx {
+            options: options.clone(),
+            output_id: output_id.clone(),
+            request_keyframe_sender,
+            should_close: should_close.clone(),
+            tokio_rt: pipeline_ctx.tokio_rt.clone(),
+        };
+
+        pipeline_ctx.tokio_rt.spawn(
+            run_whip_sender_task(
+                whip_ctx,
+                packet_stream,
+                init_confirmation_sender,
+                pipeline_ctx.event_emitter.clone(),
+            )
+            .instrument(span!(
+                Level::INFO,
+                "WHIP sender",
+                output_id = output_id.to_string()
+            )),
+        );
+
+        let start_time = Instant::now();
+        loop {
+            thread::sleep(Duration::from_millis(500));
+            let elapsed_time = Instant::now().duration_since(start_time);
+            if elapsed_time > WHIP_INIT_TIMEOUT {
+                return Err(OutputInitError::WhipInitTimeout);
+            }
+            match init_confirmation_receiver.try_recv() {
+                Ok(result) => match result {
+                    Ok(_) => break,
+                    Err(err) => return Err(OutputInitError::WhipInitError(err.into())),
+                },
+                Err(err) => match err {
+                    oneshot::error::TryRecvError::Closed => {
+                        return Err(OutputInitError::UnknownWhipError)
+                    }
+                    oneshot::error::TryRecvError::Empty => {}
+                },
+            };
+        }
+
+        Ok(Self {
+            connection_options: options,
+            should_close,
+        })
+    }
+}
+
+impl Drop for WhipSender {
+    fn drop(&mut self) {
+        self.should_close
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+async fn run_whip_sender_task(
+    whip_ctx: WhipCtx,
+    packet_stream: PacketStream,
+    init_confirmation_sender: oneshot::Sender<Result<(), WhipError>>,
+    event_emitter: Arc<EventEmitter>,
+) {
+    let client = Arc::new(reqwest::Client::new());
+    let (peer_connection, video_track, audio_track) = match init_peer_connection().await {
+        Ok(pc) => pc,
+        Err(err) => {
+            init_confirmation_sender.send(Err(err)).unwrap();
+            return;
+        }
+    };
+    let output_id_clone = whip_ctx.output_id.clone();
+    let should_close_clone = whip_ctx.should_close.clone();
+    let whip_session_url = match connect(peer_connection, client.clone(), whip_ctx).await {
```

Comment:
```suggestion
    let whip_session_url = match connect(peer_connection, client.clone(), whip_ctx.clone()).await {
```
or 
```suggestion
    let whip_session_url = match connect(peer_connection, client.clone(), &whip_ctx).await {
```
and clone ctx inside